src/v2/manifest/mod.rs

```diff
@@ -77,22 +90,68 @@ impl Client {
                 }
             })
             .and_then(|res| {
-                future::ok(res.headers().clone()).join(
+                future::ok((res.headers().clone(), res.url().clone())).join(
                     res.into_body()
                         .concat2()
                         .map_err(|e| Error::from(format!("{}", e))),
                 )
             })
             .map_err(|e| Error::from(format!("{}", e)))
-            .and_then(|(headers, body)| {
+            .and_then(|((headers, url), body)| {
                 let content_digest = match headers.get("docker-content-digest") {
-                    Some(content_digest_value) => Some(content_digest_value.to_str()?.to_string()),
+                    Some(content_digest_value) => Some(
+                        content_digest_value
+                            .to_str()
+                            .map_err(Error::from)?
+                            .to_string(),
+                    ),
                     None => {
                         debug!("cannot find manifestref in headers");
                         None
                     }
                 };
-                Ok((body.to_vec(), content_digest))
+
+                let header_content_type = headers.get(header::CONTENT_TYPE);
+                let media_type = evaluate_media_type(header_content_type, &url)?;
+
+                trace!(
+                    "content-type: {:?}, media-type: {:?}",
+                    header_content_type,
+                    media_type
+                );
+
+                Ok((body, content_digest, media_type))
+            })
+            .and_then(move |(body, content_digest, media_type)| {
+                match media_type {
+                    mediatypes::MediaTypes::ManifestV2S1Signed => {
+                        Either::A(futures::future::result(
```

Comment:
Its not clear why its different from `ManifestV2S2` code.

Shouldn't it be
```
mediatypes::MediaTypes::ManifestV2S1Signed => Either::A(
	futures::future::result(serde_json::from_slice::<ManifestSchema1Signed>(
		&body,
	))
	.map_err(|e| Error::from(format!("{}", e)))
...
```
?

Reply:
Are you wondering about the curly braces?

Reply:
Yeah, pretty much. This is probably very minor issue, just a bit confusing if these cases are similar

Reply:
The braces are added back by `cargofmt` if I remove them; that settles the styling issue.
The `ManifestV2S2` case has the additional step of downloading the config blob, which is a non-immediate future. compared to the `ManifestV2S1` case which only has the immediate value inside the future. That's why they are different as in `Either::A` and `Either::B`.